Allow provider specific examples

nagutm opened the pull request:
Closes #1417

It updates the web front-end to check if a provider-specific example is available and, if so, modifies the provider link in the Extra Providers sections to use the custom example ID instead of the default one.

As a test I added a custom example for `bfvd` which is a provider for `uniprot` and it correctly displays the link with the custom example and is able to resolve the ID appropriately.

<img width="910" alt="image" src="https://github.com/user-attachments/assets/b4bed16c-af7e-4ee0-9091-84b476305880" />

cthoyt reviewed (changes requested):
@nagutm thanks for looking into this! One thing that's important for any new schema change proposals is to add a real curation to bioregistry.json that reflects usage of the new schema element.

Specifically for this one, we want to make sure we apply the regular expression pattern to validate the ID is valid. You can use the code that checks the primary example / example extras as reference

Speaking of example extras, there should be a nice way of aggregating any possible specific examples put in the providers into the example_extras field. The question is: should we require double curation there, or should we have the function that wraps it automatically grab and aggregate them? Check usages of `.example_extras` to see what might be more comfy

nagutm commented:
> Specifically for this one, we want to make sure we apply the regular expression pattern to validate the ID is valid. You can use the code that checks the primary example / example extras as reference

@cthoyt Could you clarify exactly what piece of code you are referring to here that checks the primary example/example extras?

nagutm commented:
> Speaking of example extras, there should be a nice way of aggregating any possible specific examples put in the providers into the example_extras field. The question is: should we require double curation there, or should we have the function that wraps it automatically grab and aggregate them? Check usages of `.example_extras` to see what might be more comfy

I added a method to the Resource class called `get_aggregated_example_extras()` which concatenates the manually curated example_extras along with any provider specific examples. I also updated any references to `.example_extras` with this method to ensure that provider-specific examples are included in any list of example_extras in the future.

bgyori commented:
> > Specifically for this one, we want to make sure we apply the regular expression pattern to validate the ID is valid. You can use the code that checks the primary example / example extras as reference
> 
> @cthoyt Could you clarify exactly what piece of code you are referring to here that checks the primary example/example extras?

There are tests that check whether example identifier for a given entry conform to the identifier regular expression pattern definition. These tests should cover not only the default and extra examples but also the examples specific to providers so if an invalid example identifier is added, the test fails.

nagutm commented on tests/test_data.py:
I believe that by replacing `.example_extras` with `.get_aggregated_example_extras` this will assert that any provider specific examples are also conforming to the identifier pattern for the resource.

nagutm commented:
I added a provider example for `furna` to run some of the tests implemented in this PR. The previous link: https://seq2fun.dcmb.med.umich.edu/furna/pdb.cgi?pdbid=2gc4 was not resolving the default `pdb` ID as expected.

cthoyt commented on tests/test_data.py:
```diff
@@ -744,6 +744,13 @@ def test_providers(self):
                         )
                         self.assert_publication_identifiers(publication)
+                    # Check provider-specific example handling
+                    if provider.example:
```
Is there not already a test that checks there's a `$1` character in the provider's URI format?

I think this added block can be removed

cthoyt commented on src/bioregistry/schema/struct.py:
```diff
@@ -1353,8 +1361,21 @@ def get_examples(self) -> list[str]:
+        rv.extend(self.get_aggregated_example_extras() or [])
         return rv
+    
+    def get_aggregated_example_extras(self) -> list[str]:
```
Please rename to `get_example_extras`

cthoyt commented on src/bioregistry/schema/struct.py:
```diff
@@ -1353,8 +1361,21 @@ def get_examples(self) -> list[str]:
+                if provider.example:
+                    aggregated_examples.add(provider.example)
+
+        return list(aggregated_examples)
```
for consistency

```suggestion
        return sorted(aggregated_examples)
```

cthoyt reviewed (changes requested):
besides the two comments I left, this is ready to go. Thanks @nagutm
